tests/ui/needless_lifetimes.rs

```diff
@@ -259,4 +259,36 @@ mod issue4291 {
     }
 }
 
+mod nested_elision_sites {
+    // Don't lint these cases, they cause FPs.
+    // The lint does not support nested elision sites.
+
+    fn nested_fn_trait_bound<'a>(i: &'a i32) -> impl Fn() -> &'a i32 {
+        move || i
+    }
+
+    fn nested_fn_mut_trait_bound<'a>(i: &'a i32) -> impl FnMut() -> &'a i32 {
+        move || i
+    }
+
+    fn nested_fn_once_trait_bound<'a>(i: &'a i32) -> impl FnOnce() -> &'a i32 {
+        move || i
+    }
+
+    fn nested_generic_fn_trait_bound<'a, T: Fn() -> &'a i32>(f: T) -> &'a i32 {
+        f()
+    }
+
+    fn nested_where_clause_fn_trait_bound<'a, T>(f: T) -> &'a i32
+    where
+        T: Fn() -> &'a i32,
+    {
+        f()
+    }
```

Comment:
I think these two cases were already covered by the previous tests.

Reply:
Indeed, but for a different reason. In the implementation before this PR, the lint gives up if any named lifetime is found in the where clause (the implementation is more sophisticated than that, see `has_where_lifetimes`. I think this is a limitation of the lint and might be improved in the future. I added theses test cases as they are independent of this limitation, in order to protect against potential regressions around nested elision sites in where clauses.

Reply:
I'll reorder the tests and try to remove duplicates.

Reply:
> I added these test cases as they are independent of this limitation, in order to protect against potential regressions around nested elision sites in where clauses.

My bad, that makes sense :+1: